File: src/LoadingPanelController.m
//
//  LoadingPanelController.m
//  Disk Inventory X
//
//  Created by Tjark Derlien on 03.12.04.
//
//  Copyright (C) 2004 Tjark Derlien.
//  
//  This program is free software; you can redistribute it and/or
//  modify it under the terms of the GNU General Public License
//  as published by the Free Software Foundation; either version 3
//  of the License, or any later version.

//

#import "LoadingPanelController.h"
#import "Timing.h"


@implementation LoadingPanelController

- (id) init
{
	self = [super init];

    //load Nib with progress panel
	NSArray *topLevelObjects = nil;
	if ( ![[NSBundle mainBundle] loadNibNamed: @"LoadingPanel" owner: self topLevelObjects: &topLevelObjects] )
		NSAssert( NO, @"couldn't load LoadingPanel.nib" );
	_nibTopLevelObjects = topLevelObjects;

	//the nib has "release when closed = YES"; turn it off so the panel's
	//lifetime is governed by _nibTopLevelObjects only. Otherwise [_loadingPanel
	//close] frees the panel while the array still holds a stale pointer to it,
	//producing a zombie crash when the controller is later released.
	[_loadingPanel setReleasedWhenClosed: NO];

	[_loadingProgressIndicator setUsesThreadedAnimation: NO];
    [_loadingProgressIndicator startAnimation: self];

	[_loadingPanel display];

	//start modal session for the progress window
	_loadingPanelModalSession = [[NSApplication sharedApplication] beginModalSessionForWindow: _loadingPanel];
	_lastEventLoopRun = 0;
	
	_cancelPressed = NO;
	
	return self;
}

- (id) initAsSheetForWindow: (NSWindow*) window
{
	self = [super init];

    //load Nib with progress panel
	NSArray *topLevelObjects = nil;
	if ( ![[NSBundle mainBundle] loadNibNamed: @"LoadingPanel" owner: self topLevelObjects: &topLevelObjects] )
		NSAssert( NO, @"couldn't load LoadingPanel.nib" );
	_nibTopLevelObjects = topLevelObjects;

	//see init - same reason.
	[_loadingPanel setReleasedWhenClosed: NO];

	[window beginSheet: _loadingPanel completionHandler: nil];

	[_loadingPanel setWorksWhenModal: YES];
	
	[_loadingProgressIndicator setUsesThreadedAnimation: NO];
    [_loadingProgressIndicator startAnimation: self];
	
	//we don't have modal session if we show the panel as a sheet
	_loadingPanelModalSession = 0;
	
	_lastEventLoopRun = 0;
	
	_cancelPressed = NO;
	
	return self;
}

- (void) dealloc
{
	if ( _loadingPanel != nil )
		[self close];
}

- (void) close
{
	if ( [_loadingPanel isSheet] )
	{
		[[_loadingPanel sheetParent] endSheet: _loadingPanel];
		[_loadingPanel close]; //just hides; ownership stays with _nibTopLevelObjects
		
		_loadingPanel = nil;
		_loadingProgressIndicator = nil;
		_loadingTextField = nil;
		_loadingCancelButton = nil;
	}
	else
	{
		OBPRECONDITION( _loadingPanelModalSession != 0 );
		[[NSApplication sharedApplication] endModalSession: _loadingPanelModalSession];
		_loadingPanelModalSession = 0;
		
		[self closeNoModalEnd];
	}
}

- (void) closeNoModalEnd
{
	//this only works if we startet a modal session for a panel (no sheet)
	OBPRECONDITION( ![_loadingPanel isSheet] );
	
	//the sender asked us not to end the modal session (maybe because sender has run into an exception)
	_loadingPanelModalSession = 0;
	
	[_loadingPanel close]; //will be released (panel has style "release when close")
	
	_loadingPanel = nil;
    _loadingProgressIndicator = nil;
	_loadingTextField = nil;
	_loadingCancelButton = nil;
}

- (void) enableCancelButton: (BOOL) enable
{
	[_loadingCancelButton setEnabled: enable];
}

- (BOOL) cancelPressed
{
	return _cancelPressed;
}

- (void) startAnimation;
{
	[_loadingProgressIndicator startAnimation: nil];
}

- (void) stopAnimation;
{
	[_loadingProgressIndicator stopAnimation: nil];
}

- (void) setMessageText: (NSString*) msg
{
	_message = msg;
}

- (void) runEventLoop
{
	//we only let the UI update itself every 0.2 second, otherwise running
	//the event loop eats over half of the total scan time!
	uint64_t currentTime = getTime();
	BOOL runEventLoop = _lastEventLoopRun == 0 || subtractTime( currentTime, _lastEventLoopRun ) > 0.2;

	if ( _message != nil )
	{
		[_loadingTextField setStringValue: _message];
		
		//set message to nil so it won't be set a again in the NSTextField
		[self setMessageText: nil];
			
		//if we don't run the event loop, just update the text field
		if ( !runEventLoop )
			[_loadingTextField displayIfNeeded];
	}
	
	if ( runEventLoop )
	{
		_lastEventLoopRun = currentTime;
		
		//give progress dialog some processor cycles
		if ( _loadingPanelModalSession != 0 )
		{
			if ( [[NSApplication sharedApplication] runModalSession: _loadingPanelModalSession]
																			!= NSModalResponseContinue )
			{
				NSAssert( NO, @"run loop stopped by unknown party" );
			}
		}
		else
		{
			[[NSRunLoop currentRunLoop] runUntilDate: [NSDate date]];
		}
	}
}

- (IBAction) cancel:(id)sender
{
	_cancelPressed = YES;
	
	[_loadingCancelButton setEnabled: NO];
}

@end

File: src/LoadingPanelController.swift
//
//  LoadingPanelController.swift
//  Disk Inventory Xs
//
//  Swift port of LoadingPanelController.{h,m}. Owns LoadingPanel.nib and drives
//  the scan progress panel (either modal or as a sheet), throttling event-loop
//  pumping to ~5 Hz so it doesn't dominate scan time. Nib owner, so the class
//  name, IBOutlets, and the cancel: action are pinned with @objc.
//
//  GPL v3
//

import Cocoa

@objc(LoadingPanelController)
class LoadingPanelController: NSObject {

    @IBOutlet private var _loadingTextField: NSTextField!
    @IBOutlet private var _loadingPanel: NSPanel!
    @IBOutlet private var _loadingProgressIndicator: NSProgressIndicator!
    @IBOutlet private var _loadingCancelButton: NSButton!

    private var _loadingPanelModalSession: NSApplication.ModalSession?
    private var _lastEventLoopRun: UInt64 = 0
    private var _cancelPressed = false
    private var _message: String?
    private var _nibTopLevelObjects: [Any]?

    // starts a modal session immediately
    @objc override init() {
        super.init()
        loadPanelNib()

        // the nib has "release when closed = YES"; turn it off so the panel's
        // lifetime is governed by _nibTopLevelObjects only. Otherwise closing it
        // frees the panel while the array still holds a stale pointer, producing
        // a zombie crash when the controller is later released.
        _loadingPanel.isReleasedWhenClosed = false

        _loadingProgressIndicator.usesThreadedAnimation = false
        _loadingProgressIndicator.startAnimation(self)
        _loadingPanel.display()

        _loadingPanelModalSession = NSApplication.shared.beginModalSession(for: _loadingPanel)
        _lastEventLoopRun = 0
        _cancelPressed = false
    }

    // starts the panel as a sheet immediately
    @objc(initAsSheetForWindow:) init(asSheetFor window: NSWindow) {
        super.init()
        loadPanelNib()

        _loadingPanel.isReleasedWhenClosed = false  // see init() for the reason

        window.beginSheet(_loadingPanel, completionHandler: nil)
        _loadingPanel.worksWhenModal = true

        _loadingProgressIndicator.usesThreadedAnimation = false
        _loadingProgressIndicator.startAnimation(self)

        // no modal session when shown as a sheet
        _loadingPanelModalSession = nil
        _lastEventLoopRun = 0
        _cancelPressed = false
    }

    private func loadPanelNib() {
        var topLevelObjects: NSArray?
        if !Bundle.main.loadNibNamed("LoadingPanel", owner: self, topLevelObjects: &topLevelObjects) {
            assertionFailure("couldn't load LoadingPanel.nib")
        }
        _nibTopLevelObjects = topLevelObjects as? [Any]
    }

    deinit {
        if _loadingPanel != nil {
            close()
        }
    }

    @objc func close() {
        if _loadingPanel.isSheet {
            _loadingPanel.sheetParent?.endSheet(_loadingPanel)
            _loadingPanel.close()  // just hides; ownership stays with _nibTopLevelObjects
        } else {
            if let session = _loadingPanelModalSession {
                NSApplication.shared.endModalSession(session)
            }
            _loadingPanelModalSession = nil
            closeNoModalEnd()
            return
        }
        _loadingPanel = nil
        _loadingProgressIndicator = nil
        _loadingTextField = nil
        _loadingCancelButton = nil
    }

    @objc func closeNoModalEnd() {
        // the sender asked us not to end the modal session (e.g. it hit an exception)
        _loadingPanelModalSession = nil
        _loadingPanel?.close()
        _loadingPanel = nil
        _loadingProgressIndicator = nil
        _loadingTextField = nil
        _loadingCancelButton = nil
    }

    @objc(enableCancelButton:) func enableCancelButton(_ enable: Bool) {
        _loadingCancelButton.isEnabled = enable
    }

    @objc func cancelPressed() -> Bool { _cancelPressed }

    @objc func startAnimation() { _loadingProgressIndicator.startAnimation(nil) }
    @objc func stopAnimation() { _loadingProgressIndicator.stopAnimation(nil) }

    // message will be shown next time runEventLoop is called
    @objc(setMessageText:) func setMessageText(_ msg: String?) { _message = msg }

    @objc func runEventLoop() {
        // only refresh the UI every 0.2s; pumping the event loop more often eats
        // over half the total scan time
        let currentTime = getTime()
        let pumpEventLoop = _lastEventLoopRun == 0 || subtractTime(currentTime, _lastEventLoopRun) > 0.2

        if let message = _message {
            _loadingTextField.stringValue = message
            setMessageText(nil)  // so it isn't reapplied next time
            if !pumpEventLoop {
                _loadingTextField.displayIfNeeded()
            }
        }

        if pumpEventLoop {
            _lastEventLoopRun = currentTime
            if let session = _loadingPanelModalSession {
                if NSApplication.shared.runModalSession(session) != .continue {
                    assertionFailure("run loop stopped by unknown party")
                }
            } else {
                RunLoop.current.run(until: Date())
            }
        }
    }

    @IBAction func cancel(_ sender: Any?) {
        _cancelPressed = true
        _loadingCancelButton.isEnabled = false
    }
}
